### ISSUE TITLE
SIGSEGV when writing more than blockSize

### ISSUE THREAD

**Issue by handmadecode** [OpenHFT/Chronicle-Queue#561]
I'm getting a SIGSEGV when writing to a chronicle queue using version 5.17.3. This does not seem to happen with versions 4.5.15 and 4.15.7.

Problematic frame: # v  ~StubRoutines::jlong_disjoint_arraycopy

The problem occurs when writing more than the block size to the queue. It doesn't happen every time, but approximately 80% of my test runs crash.

Tested with 
    java version "1.8.0_192"
    Java(TM) SE Runtime Environment (build 1.8.0_192-b12)
    Java HotSpot(TM) 64-Bit Server VM (build 25.192-b12, mixed mode)
and
    openjdk version "11.0.1" 2018-10-16
    OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.1+13)
    OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.1+13, mixed mode)
on macOS 10.14.2

The following snippet (also attached as a test case) illustrates the problem:

```
        Path aPath = Files.createTempDirectory("ChronicleTest");
        byte[] aBytes = new byte[512];
        ByteBuffer aByteBuffer = ByteBuffer.wrap(aBytes);

        try (SingleChronicleQueue aQueue = SingleChronicleQueueBuilder.binary(fChronicleDirectory).build())
        {
            ExcerptAppender aAppender = aQueue.acquireAppender();

            long aNumBytesWritten = 0, aBlockSize = aQueue.blockSize();
            while (aNumBytesWritten <= 10 * aBlockSize)
            {
                ThreadLocalRandom.current().nextBytes(aBytes);
                aByteBuffer.clear();
                aAppender.writeBytes(b -> b.writeSome(aByteBuffer));
                aNumBytesWritten += aByteBuffer.capacity() - aByteBuffer.remaining();
            }
        }
```

[CrashTest.java.txt](https://github.com/OpenHFT/Chronicle-Queue/files/2782563/CrashTest.java.txt)

**Comment by MartyIX** [comment]
@handmadecode I believe you can change 

```
b -> b.writeSome(aByteBuffer)
```

to

```
b -> b.write(bytes, 0, bytes.length);
```

`writeSome` leads to calling `memory.copyMemory(...)` but as `writeSome` method does not check boundaries (see https://github.com/OpenHFT/Chronicle-Bytes/blob/master/src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java#L890) it seems to corrupt memory.

On the other  hand, `write(@NotNull byte[] bytes, int offset, int length)` (https://github.com/OpenHFT/Chronicle-Bytes/blob/master/src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java#L868) does verify `writeRemaining()` limit and as such it should not fail (it took too long on my computer so I stopped the test eventually).

**Comment by handmadecode** [comment]
@MartyIX Thanks for looking into this.

Unfortunately I cannot use `write(@NotNull byte[] bytes, int offset, int length)` without creating an intermediate `byte[]`, the code that writes to the Chronicle Queue is handed the ByteBuffer by its clients, and it will most likely be a direct buffer.

Is there any other way of writing the contents of a ByteBuffer without copying the bytes to some other data structure first?

Also, I have continued using version 4.15.7 of Chronicle Queue, and still haven't seen the problem with that version. As far as I can tell the `writeSome` method hasn't changed between that version and master.

**Comment by MartyIX** [comment]
@handmadecode Sorry, my analysis was flawed. 

However, there are closed issues that seems similar: https://github.com/OpenHFT/Chronicle-Queue/issues?utf8=%E2%9C%93&q=is%3Aissue+blockSize - mainly:

* https://github.com/OpenHFT/Chronicle-Queue/issues/487
* https://github.com/OpenHFT/Chronicle-Queue/issues/535

Based on that, it seems that you write to some random memory and that leads to the crash. Solution is to increase your blockSize, if possible.

**Comment by handmadecode** [comment]
Yes, I had a look at those issues and they do seem related. However, the problem I encountered is not caused by writing a single large entry to the queue, but rather by writing many small entries that when they add up to more than the blockSize cause the SIGSEGV.

I've successfully written several millions of small entries (100-200 bytes each) adding up to about ten times the default block size, using Chronicle Queue 4.15 (and chronicle-bytes 1.15.4), which I don't think has that fix.

Since everything seems to work with Chronicle Queue 4.15 this is not a blocker for me. I'll see if I can isolate the problem more in the coming week.

**Comment by dpisklov** [comment]
@handmadecode thanks for raising this. Apparently this has uncovered a pretty serious issue - because of incorrect range check we ended up overwriting the queue file header (hence the SIGSEGV - when reading header we read some garbage written on previous iteration). I have fixed it now.

**Comment by handmadecode** [comment]
Good news, thanks for fixing this.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.5](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.5), [BOM-2.17.196](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.196)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.5](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.5), [BOM-2.17.225](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.225)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.5](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.5), [BOM-2.17.241](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.241)
